pkg/machinery/config/decoder/decoder.go

```diff
@@ -164,5 +171,97 @@ func decode(manifest *yaml.Node) (target interface{}, err error) {
 		return nil, fmt.Errorf("spec decode: %w", err)
 	}
 
+	if err = validate(target, spec); err != nil {
+		return nil, err
+	}
+
 	return target, nil
 }
+
+//nolint:gocyclo
+func validate(target interface{}, spec *yaml.Node) error {
+	node, err := encoder.NewEncoder(target).Marshal()
+	if err != nil {
+		return err
+	}
+
+	src := map[string]interface{}{}
+	dst := map[string]interface{}{}
+
+	err = spec.Decode(src)
+	if err != nil {
+		return err
+	}
+
+	err = node.Decode(dst)
+	if err != nil {
+		return err
+	}
+
+	var errs *multierror.Error
+
+	keys := []string{}
+
+	var checkExtra func(interface{}, interface{})
+
+	checkExtra = func(left interface{}, right interface{}) {
+		switch v := left.(type) {
+		case map[string]interface{}:
+			r, ok := right.(map[string]interface{})
+			if !ok {
+				errs = multierror.Append(errs, fmt.Errorf("type mismatch %v, expected map", strings.Join(keys, ".")))
+
+				return
+			}
+
+			for key, value := range v {
+				if _, ok := r[key]; !ok {
+					errs = multierror.Append(errs, fmt.Errorf("extra key %v", strings.Join(append(keys, key), ".")))
+
+					continue
+				}
+
+				func() {
```

Comment:
this works, but would it be a bit easier to pass `keys` as an argument to `checkExtra`, that might be called `path` ?

Reply:
yeah, that's another option